### PR TITLE
Reactoring into single-process, multi-thread

### PR DESCRIPTION
**Pull request by facontidavide**
For your information (I am aware that there is a merge conflict) here you can find my refactoring that natively load a rosbag and process it at the maximum speed allowed by your CPU.

**This PR is not meant to be merged** but is my way to share with you our work and start discussing how such merge can be done in the future, if you want to.

The nice thing about this version is that:

- it has the same level of parallelism as the original code, but...
- it is deterministic, i.e. given the same rosbag, the result will always be the same.
- inter-thread communication is much more lightweight then the ROS transport.

The determinism allows me to introduce more changes with confidence, because I can alwasys double check if a certain change introduced a change in terms of accuracy or not.

For instance, there is another branch where I made the whole application considerably faster:
https://github.com/facontidavide/LeGO-LOAM/tree/speed_optimization

### CODE REVIEW

File: LeGO-LOAM/CMakeLists.txt [hunk 1]
```diff
@@ -45,16 +48,13 @@ link_directories(
...
 )
 
-add_executable(imageProjection src/imageProjection.cpp)
-add_dependencies(imageProjection ${catkin_EXPORTED_TARGETS} cloud_msgs_gencpp)
-target_link_libraries(imageProjection ${catkin_LIBRARIES} ${PCL_LIBRARIES} ${OpenCV_LIBRARIES})
-
-add_executable(featureAssociation src/featureAssociation.cpp)
-add_dependencies(featureAssociation ${catkin_EXPORTED_TARGETS} cloud_msgs_gencpp)
-target_link_libraries(featureAssociation ${catkin_LIBRARIES} ${PCL_LIBRARIES} ${OpenCV_LIBRARIES})
+add_executable(lego_loam
```
[the remark this answers is not in the document]
Comment (facontidavide) [reply LeGO-LOAM/CMakeLists.txt]: Ok

**Comment by facontidavide** [comment]
Sure, that would be trivial. But I prefer to do it in this way, because either you have sufficient throughput or you don't. If the reading thread isn't fast enough, the buffer will grow anyway

File: LeGO-LOAM/CMakeLists.txt [hunk 2]
```diff
@@ -1,14 +1,17 @@
...
+
+#
 
 find_package(catkin REQUIRED COMPONENTS
   tf
   roscpp
   rospy
   cv_bridge
   image_transport
+  rosbag_storage
```
[the remark this answers is not in the document]
Comment (facontidavide) [reply LeGO-LOAM/CMakeLists.txt]: Rosbag is the command line app. Rosbag_storage is the lower level c++ library

[the remark this answers is not in the document]
Comment (andre-nguyen) [reply LeGO-LOAM/CMakeLists.txt]: Interesting, I had always used the `rosbag` target as in here 

https://github.com/davetcoleman/rosbag_record_cpp/blob/4de3c5b1199f8e3edd6194989c43e47d9b0ea1b8/CMakeLists.txt#L7-L11

**Pull request closed without merging**